File: trees/ltcd_2415.py
'''
LeetCode - problem 2415
'''
    from collections import defaultdict

    def reverseOddLevels(self, root):
        odd_level_nodes = defaultdict(list)

        def reverse(root, level=0, side=None):
            if root is None:
                return None

            if level % 2 == 1:
                if side == "right":
                    odd_level_nodes[level].append(root)
                else: 
                    mirror_root = odd_level_nodes[level].pop()
                    root.val, mirror_root.val = mirror_root.val, root.val

            root.right = reverse(root.right, level+1, "right" if side is None else side)
            root.left = reverse(root.left, level+1, "left" if side is None else side)
            return root

        return reverse(root)

File: trees/ltcd_2415.cpp
//LeetCode - problem 2415

TreeNode* reverseOddLevels(TreeNode* root) {
	vector<TreeNode*> v;
	v.push_back(root);
	bool isOdd = true;
	while (v.size() != 0) {
		vector<TreeNode*> temp;
		for (TreeNode* n : v) {
			if (n->left) temp.push_back(n->left);
			if (n->right) temp.push_back(n->right);
		}
		if (isOdd) {
			int front = 0, back = temp.size() - 1;
			while (front < back) {
				swap(temp[front]->val, temp[back]->val);
				front++;
				back--;
			}
		}
		v = temp;
		isOdd = !isOdd;
	}
	return root;
}
